src/nvme/util.c

```diff
@@ -403,14 +403,17 @@ void nvme_init_copy_range_f1(struct nvme_copy_range_f1 *copy, __u16 *nlbs,
 			  __u64 *slbas, __u64 *eilbrts, __u32 *elbatms,
 			  __u32 *elbats, __u16 nr)
 {
-	int i;
+	int i, j;
 
 	for (i = 0; i < nr; i++) {
 		copy[i].nlb = cpu_to_le16(nlbs[i]);
 		copy[i].slba = cpu_to_le64(slbas[i]);
-		copy[i].elbt[2] = cpu_to_le64(eilbrts[i]);
 		copy[i].elbatm = cpu_to_le16(elbatms[i]);
 		copy[i].elbat = cpu_to_le16(elbats[i]);
+		for (j = 0; j < 8; j++)
+			copy[i].elbt[9 - j] = (eilbrts[i] >> (8 * j)) & 0xff;
```

Comment:
I think the issue here is that the EILBRT values are stored in big-endian order, like the linked PR which was closed.

Storing them in little-endian order requires the Storage Tag Size and Protection Information Format to determine where the EILBRT should start in the shared range.

Reply:
Since both EILBRT and ELBST values are store in  big-endian order, and the device software know the Storage Tag Size and what kind of Protection Information Format used, so i think the location of  EILBRT in the shared range is handle by device software, as is the ELBST.
For nvme-cli，we only need to deliver EILBRT and ELBST  in big-end mode to the device.

Reply:
> Since both EILBRT and ELBST values are store in big-endian order

Most of the spec is using little endian and I can't find anything EILBRT and ELBST being big endian. The EILBRT and ELBST definition is, let's say a 'challenging' definition. Where do you find the info on big-endian?

Reply:
![image](https://user-images.githubusercontent.com/18096786/223312478-e18f1594-7540-4c40-a970-b0ef2094d740.png)
 in section 5.2.1 Protection Information Formats. for example, the screenshot show the 16b guard protection information format, EILBRT and ELBST re store in big-endian order

Reply:
Ah, I didn't look at the Figure :)

I think you should use the be64 helpers as the argument type for eilbert is u64 which is in host endian.

But as I said pointed out, this change breaks the API. We can't just change data structures in this way. This was the reason why I didn't merge #455.

Reply:
Yeah I'd thought that the MSB/LSB in that figure might be "Most/Least Significant Bits" rather than defining each of these fields as big-endian byte ordering, particularly based on figure 122 in the next section on the shared space and the consistent mention of significant bits rather than bytes:

![image](https://user-images.githubusercontent.com/24419498/223832883-9ff361e6-e523-4185-9034-5b2c058bc53c.png)

Definitely see how it could be describing the byte-ordering, but it feels odd without a more explicit call-out like NGUID or EUI64. Worth mentioning this would mean similar ordering changes are likely needed for the shared field and for application tag/guard fields across all IO commands.

Reply:
Thanks for the feedback. I looked again with fresh eyes I figured that the github webui confused me. So the two commits should be obviously merged and the final result is just:

```diff
--- a/src/nvme/types.h
+++ b/src/nvme/types.h
@@ -4640,8 +4640,8 @@ struct nvme_copy_range {
        __le16                  nlb;
        __u8                    rsvd18[6];
        __le32                  eilbrt;
-       __le16                  elbatm;
        __le16                  elbat;
+       __le16                  elbatm;
 };
 
 /**
@@ -4661,8 +4661,8 @@ struct nvme_copy_range_f1 {
        __le16                  nlb;
        __u8                    rsvd18[8];
        __u8                    elbt[10];
-       __le16                  elbatm;
        __le16                  elbat;
+       __le16                  elbatm;
 };
```

which is okay I think to merge.